Add support for ignoring package source location (fixes issue #1)

torkildr opened the pull request:
This commit adds support for ignoring package source location, using --no-from.

Additionally, I've upped the version to 0.0.4 (from the npm package), seeing as the master repo seems to contain old code.

torkildr committed:
Add support for ignoring package source location

dsagal commented:
Thank you Torkild! Firstly, apologies for forgetting to push to github -- I did now, and it corresponds to npmjs. So I suggest you rebase.

This is to solve Issue #1, which is about a missing or mismatching "from" field. I want to make sure you know: if you run `npm shrinkwrap --dev`, it should fix `npm-shrinkwrap.json` so that "from" fields correspond. I'd like to understand when it's not appropriate.

Also, I tested and noticed that if you have a MISSING _from, then your pull request doesn't seem to help.

torkildr commented:
Thank you for fast response, and for updating the repo!

My use case is something I've come upon in a team setting. The scenario is that I'm upping the version on some package, and I want others on the team to be able to verify that they've got the correct set of packages installed (_without_ running npm install). I'd like to be able not to touch the npm-shrinkwrap file.

In this specific case, a member of our team got a false negative, saying that a package was incorrect, because his package location did not correspond to the one in the npm-shrinkwrap-file, even though the version was correct.

I'll take a look into rebasing on the updated branch. And I'll see if I can make test case out of the missing "_from", and see if I can make sense of that as well.

dsagal commented:
Interesting. The most common scenario is when both npm-shrinkwrap and package.json contain just a version for "from" (or version range). In our case, we actually use a couple of forked packages, which don't differ in version number, so their source is how we can tell if they are right.

But yes, I agree it would be nice to have a flag to silence such differences.

torkildr committed:
Add a test case where _from is missing from package.json. Make --no-from handle this case.

torkildr commented:
I have now rebased on the updated master branch.

I have also added a test case where _from is missing from the target package.json. As you said, this case was not solved by my patch, so I have restructured it in such a way that it also works for this scenario.

As far as I can see, the --no-from option shouldn't interfere anything with regular usage.

torkildr committed:
Document usage of --no-from

dsagal commented on lib/main.js:
```diff
@@ -84,7 +89,8 @@ function main(options) {
     let module = JSON.parse(data);
     for (let name in module.dependencies) {
       let values = module.dependencies[name];
-      getInfo(allModules, name).want = getInstallSpec(log, name, values.version, values.from);
```
I think if you use `let from = options.from ? values.from : null;` then getInstallSpec() will do the right thing without needing an extra argument. (I.e. it's already doing what you want when `from` argument is falsy.)

torkildr replied:
Thank you. I'd also like to compliment you on the logical coverage in the test cases. They made it very easy to expand the functionality without having too much details on the inner workings. 

I agree. I kinda started off with wanting to keep this functionality as separate as possible, but I agree that this will keep this cleaner and more readable.

dsagal reviewed (approved):
Beautiful! Thank you for the extra test case. I have a minor suggestion that I think would simplify the code a little: could you check and tell me if you agree?

torkildr committed:
Simplify --no-from logic.
